pkg/scheduler/util/utils_test.go

```diff
@@ -93,3 +93,209 @@ func TestSortableList(t *testing.T) {
 		}
 	}
 }
+
+type hostPortInfoParam struct {
+	protocol, ip string
+	port         int32
+}
+
+func TestHostPortInfo_AddRemove(t *testing.T) {
+	tests := []struct {
+		desc    string
+		added   []hostPortInfoParam
+		removed []hostPortInfoParam
+		length  int
+	}{
+		{
+			desc: "normal add case",
+			added: []hostPortInfoParam{
+				{"TCP", "127.0.0.1", 79},
+				{"UDP", "127.0.0.1", 80},
+				{"TCP", "127.0.0.1", 81},
+				{"TCP", "127.0.0.1", 82},
+				// this might not make sense in real case, but the struct doesn't forbid it.
+				{"TCP", "0.0.0.0", 79},
+				{"UDP", "0.0.0.0", 80},
+				{"TCP", "0.0.0.0", 81},
+				{"TCP", "0.0.0.0", 82},
+			},
+			length: 8,
+		},
+		{
+			desc: "empty ip and protocol add should work",
+			added: []hostPortInfoParam{
+				{"", "127.0.0.1", 79},
+				{"UDP", "127.0.0.1", 80},
+				{"", "127.0.0.1", 81},
+				{"", "127.0.0.1", 82},
+				{"", "", 79},
+				{"UDP", "", 80},
+				{"", "", 81},
+				{"", "", 82},
```

Comment:
Just out of curiosity, what will happen if someone provides a negative port number?

Reply:
My question is not necessarily related to this line of code. It is a general question.

Reply:
scheduler will reserve the negative port as well. however, the api ensures the ports are always positive, so it should be good. but you are right, scheduler can be extra cautious about this kind of input. 

```golang
	// Number of port to expose on the host.
	// If specified, this must be a valid port number, 0 < x < 65536.
	// If HostNetwork is specified, this must match ContainerPort.
	// Most containers do not need this.
	// +optional
	HostPort int32 `json:"hostPort,omitempty" protobuf:"varint,2,opt,name=hostPort"`
	// Number of port to expose on the pod's IP address.
	// This must be a valid port number, 0 < x < 65536.
	ContainerPort int32 `json:"containerPort" protobuf:"varint,3,opt,name=containerPort"`
```
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L1742-L1750

Reply:
You are right. K8s API ensure that the port number is valid:
https://github.com/bsalamat/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L201

So, I think we don't need to repeat the validation in scheduler code.